### DISCUSSION TITLE
How to make shadow relative to image size

### DISCUSSION THREAD

**Discussion by ScribbleGhost** [ImageMagick/ImageMagick#3341] [Help]
Hi. I am new to ImageMagick and I don't really see any good documentation on adding shadows.

I use this command to add shadows to screenshots:

```
magick image.jpg ( +clone -background black -shadow 10x100 ) +swap -background none -layers merge +repage image-edited.png
```

The problem is that the shadow size is not relative to the image dimensions. So if I were to use this in a script, all images regardless of size will get the same shadow size. That means small images will get a huge shadow. Likewise, large images will get a tiny shadow.

Ideally, the shadow size would be relative to the image dimensions and specified in percentage.

Is there a way to achieve this?

___

Here are some examples:

Here is the original image (2400x1806px)

![image](https://user-images.githubusercontent.com/49068170/109682313-97ee3a80-7b7e-11eb-9691-3bb484746469.jpg)

Here is the same image with the shadow (using the command above):

<img width="1400" alt="image-edited" src="https://user-images.githubusercontent.com/49068170/109682430-b5230900-7b7e-11eb-9c1a-4ebf8c8d1db0.png">

Here is the original image at 400px wide with the same effect:

![image-edited-400](https://user-images.githubusercontent.com/49068170/109682727-f9aea480-7b7e-11eb-9c66-faaa9ed0297d.png)

**Comment by snibgo** [comment]
See https://imagemagick.org/script/command-line-options.php#shadow . Instead of a fixed sigma, in pixels (eg 100), you can use an "fx" expression, eg "%[fx:w*0.05]".

**Reply by fmw42** [reply]
Note that works for IM 7. It does not work in IM 6.

**Reply by ScribbleGhost** [reply]
Thanks a bunch for a super quick reply. I am glad to see it can be done 😀

In your example `w` stands for width, I assume. Does that mean the shadow size is relative to the image width? But what about the height? Is it possible to let the shadow height be relative to the height of the image, while the shadow width is relative to the width of the image?

**Comment by snibgo** [comment]
There is only one sigma, not two separate sigmas for the x and y directions. Internally, "-shadow" is built from lower-level operations, including "-blur". You could build your own shadow using two blurs, for the x and y directions.

Yes, "w" is the width in pixels. You could use "h" for the height. See https://imagemagick.org/script/escape.php for other possibilities.

**Comment by fmw42** [comment]
See documentation. It has X and Y offsets that you can control separately

`-shadow percent-opacity{xsigma}{+-}x{+-}y{%}
`Simulate an image shadow.

**Reply by ScribbleGhost** [reply]
Referring to the documentation doesn't help much when the documentation says basically nothing about the issue at hand.

This is all it says about shadows in the documentation as far as I can tell:

> -shadow percent-opacity{xsigma}{+-}x{+-}y{%}
> Simulate an image shadow.

Really...? It's that easy huh?

This is absolutely the most horrible documentation I have ever read.

Let's make an attempt to decode this mess...

- Am I supposed to know how to write percentages? Let's say I want 50%. Should I write `50%`, `0.5`, `50`? How am I supposed to know this?
- The following hyphen after the percent, what does it mean? Does it mean that percentage is a separate value from the opacity value? Should the hyphen be included?
- The documentation repeatedly mentions xsigma. What does Xsigma mean? Where is this explained?
- `}{+-}x{+-}y{%}`.... WTH? Is this supposed to tell me something?

To be honest, I did make an effort to read the documentation, but it is very discouraging. I either feel like an idiot or that the author is an arrogant elitist. I don't want to be rude here, but the documentation really assumes that the reader can read minds.

The only reason I even managed to write the code in the original post was because I found an example online. Without it, I would be helpless 😂

**Comment by snibgo** [comment]
> Does that mean the shadow size ...

The _size_ is determined by the sigma. As fmw42 says, you can also adjust the _position_ of the shadow.

**Reply by fmw42** [reply]
I think the sigma controls the roll-off and the X and Y control the extent and angle

**Comment by snibgo** [comment]
For detail on "-shadow", with many examples, see https://legacy.imagemagick.org/Usage/blur/#shadow

**Comment by fmw42** [comment]
For novices, see

http://www.imagemagick.org/discourse-server/viewtopic.php?f=1&t=9620
http://www.imagemagick.org/script/command-line-processing.php
https://imagemagick.org/script/command-line-options.php
http://www.imagemagick.org/Usage/reference.html
http://www.imagemagick.org/Usage/
https://github.com/ImageMagick/usage-markdown
https://imagemagick.org/script/porting.php#cli

**Comment by bitflame** [comment]
Thank you so much for this. This is one of the reasons I subscribed to this list.
